Feature: BB-213 fix oplog populator race condition

welansari opened the pull request:
A race condition occurs when multiple events are processed at the same time
and that affect the same connector, in which case the connector is not configured
correctly.

For example when two buckets are being added to the same connector at the same time,
only one of them gets added, this is caused by the two operations getting and updating
the list of buckets assigned to the connector at the same time, which results in one of
them overwriting the other.

The solution is using a scheduler that periodically updates all of the connectors.
Connectors are no longer updated on an event basis

bert-e commented:
# Hello kerkesni,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/backbeat/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-213 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.8`


Please check the `Fix Version/s` of BB-213, or the target
branch of this pull request.

Commit:
feature: BB-213 fix kafka-connect connectors not being configured correctly

A race condition occurs when multiple events are processed at the same time
and that affect the same connector, in which case the connector is not configured
correctly.

For example when two buckets are being added to the same connector at the same time,
only one of them gets added, this is caused by the two operations getting and updating
the list of buckets assigned to the connector at the same time, which results in one of
them overwriting the other.

The solution is using a scheduler that periodically updates all of the connectors.
Connectors are no longer updated on an event basis

francoisferrand commented on extensions/oplogPopulator/OplogPopulatorConfigValidator.js:
```diff
@@ -8,6 +8,7 @@ const joiSchema = joi.object({
     numberOfConnectors: joi.number().required().min(1),
     prefix: joi.string().optional(),
     probeServer: probeServerJoi.default(),
+    connectorsUpdateCronRule: joi.string().default('*/5 * * * * *'),
```
every 5 minutes is a long time : we may miss lots of events  :-/

From what i understand, the problem is that we get the list of connectors every time : just getting the list once on startup (before listening to mongo events) should fix the problem as well, without introducing 5 minute delay?

Then, we may introduce a _short_ delay, to avoid too many reconfig : like once we get an update, wait a few seconds before applying it (so that we apply everything that happened since)...

welansari replied:
I actually used the same value used for all backbeat recurring tasks 5 seconds.

The issue is not in getting the list of connectors but the list of buckets assigned to a connector, this can not be done once as the values change with each update.

So with each new bucket to add (or remove) from a connector we have to get that list of buckets and update it.

When having multiple updates at the same time, the update tasks might get the wrong list (not updated), add the task's bucket and then configure the connector. This can lead to not taking in consideration the other updates.

Commit:
feature: BB-213 improve logging in oplogPopulator

Commit:
feature: BB-213 update oplogPopulator unit tests

bert-e commented:
# Branches have diverged



This pull request's source branch `feature/BB-213-fix-oplog-populator-bucket-adding-race-condition` has diverged from
`development/8.4` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/8.4` into `feature/BB-213-fix-oplog-populator-bucket-adding-race-condition`
* Rebase `feature/BB-213-fix-oplog-populator-bucket-adding-race-condition` onto `origin/development/8.4`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-213 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.16`


Please check the `Fix Version/s` of BB-213, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

jbertran commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -245,6 +249,35 @@ class ConnectorsManager {
+            const connectorsStatus = {};
+            await async.eachLimit(this._connectors, 10, async connector => {
+                try {
+                    connectorsStatus[connector.name] = {
```
This does not have to be inside the `try` block.

jbertran commented on extensions/oplogPopulator/modules/ConnectorsManager.js:
```diff
@@ -245,6 +249,35 @@ class ConnectorsManager {
+                    });
+                }
+            });
+            this._logger.info('Successfully updated connectors', {
```
This output is a little bit misleading despite the dissimilar log levels.

Can we change the message according to whether all connector updates succeeded, or connectors partially succeeded?

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers






The following reviewers are expecting changes from the author, or must review again:

* @jbertran

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

williamlardier reviewed (commented):
If I understand correctly, while I understand that removing the event-based update of the connector bucket list will help us dealing with concurrent updates/race conditions, I am not 100% sure if it is acceptable to "ignore" changes every 5s for 5s as it may lead to data loss or unwanted writes, if we're unlucky (and we will, in production!).

So I was wondering if a _stream_ approach would help us keeping the events in the right order, as they arrive, and avoid concurrent requests overriding each other? Basically, anything that could converge towards a (100%) safe solution, is this possible?

welansari commented:
@williamlardier At a second thought i do agree that even 5 seconds, is a long time for the kind of updates we are doing.
A stream based system won't solve the concurrency issue as we are still dealing with async event handlers, i believe we need controlled connector updates.

What i did for now is reduce the delay between updates to one second, but this time we only update the connectors when there are actual changes made to them (buckets added or removed), this should help reduce network calls.

williamlardier commented on extensions/oplogPopulator/modules/Connector.js:
```diff
@@ -109,13 +121,14 @@ class Connector {
      */
-    async addBucket(bucket, doUpdate = true) {
+    async addBucket(bucket, doUpdate = false) {
+        this._buckets.add(bucket);
+        this._updateConnectorState(true);
```
What happen if the `updatePipeline` fails? do we still add the bucket to the list, as it is done before the try/catch?
(I'm curious to know if the approach will be stable in case of errors, so we don't end up in an inconsistent state, but I may lack a bit of context on this one)

welansari replied:
Yes the bucket is still added to the list but `_state.bucketsGotModified` stays as true, which will trigger the next update of the connector, so we will basically be retrying until it succeeds. We also get an error log each time a connector fails

welansari replied:
i added this https://github.com/scality/backbeat/compare/17128feaebde3bc87cd6fe8afe642ba5063f0702..a0941c5e9287ea76513da7b7e7104476a3c38bc9 just to be sure the retry gets triggered

williamlardier commented on extensions/oplogPopulator/modules/Connector.js:
```diff
@@ -170,19 +184,54 @@ class Connector {
+     * connector update operation (only needed when updated connector)
+     * @returns {undefined}
+     */
+    _updateConnectorState(bucketsGotModified, timeBeforeUpdate = null) {
```
Nit:
```suggestion
    _updateConnectorState(bucketsGotModified, timeBeforeUpdate = 0) {
```

to be consistent with the jsdoc type

williamlardier reviewed (commented):
Thanks for the update, I think it's a better first step until controlled connector updates, maybe. Please find one question and one suggested change before I approve

williamlardier reviewed (approved):
LGTM, but might be better to wait for senior reviews on this one :slightly_smiling_face:

welansari commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.4`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/8.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.4`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/8.3`



Please check the status of the associated issue BB-213.

Goodbye kerkesni.
